### PR TITLE
Calculate usage per version - if statements in usage calcs

### PR DESCRIPTION
**Pull request by inaod568**
# Description

Billing computations. Get versions from database and calculate number of bytehours it has been stored since uploaded or last biilling computation. 

- [ ] Summary of the changes and the related issue
- [x] Relevant motivation and context
- [ ] Any dependencies that are required for this change

Fixes DDS-1325 and fixes DDS-1166

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# Checklist:

Please delete options that are not relevant.

- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] Rebase/merge the branch which this PR is made to
- [ ] Changes to the database schema: A new migration is included in the PR
- [ ] Product Owner / Scrum Master: This PR is made to the `master` branch and I have updated the [version](../dds_web/version.py)
- [ ] I am bumping the major version (e.g. 1.x.x to 2.x.x) and I have made the corresponding changes to the CLI version

## Formatting and documentation

- [x] I have added a row in the [changelog](../CHANGELOG.md)
- [x] The code follows the style guidelines of this project: Black / Prettier formatting
- [x] I have performed a self-review of my code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings

## Tests

- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

<a href="https://gitpod.io/#https://github.com/ScilifelabDataCentre/dds_web/pull/1246"></a>

### CODE REVIEW

File: dds_web/errors.py [hunk 1]
```diff
@@ -27,6 +27,13 @@
 extra_info = {"result": "DENIED"}
 
 
+class CronJobError(Exception):
+    """Exception notifying that something in cronjob has gone wrong."""
+
+    def __init__(self, *args: object) -> None:
+        super().__init__(*args)
+
+
```
Comment (inaod568) [inline dds_web/errors.py:36]:
```suggestion
```

File: dds_web/utils.py [hunk 2]
```diff
@@ -501,3 +502,144 @@ def get_active_motds():
...
+        print(f"- {version}", flush=True)
+        # Version uploaded after last usage calculation
+        now = current_time()
+        bytehours = calculate_bytehours(
+            minuend=now, subtrahend=version.time_uploaded, size_bytes=version.size_stored
+        )
+        project_byte_hours += bytehours
+        # version.time_invoiced = now
+
+    return project_byte_hours
```
Comment (inaod568) [inline dds_web/utils.py:645]:
```suggestion
```

File: dds_web/utils.py [hunk 3]
```diff
@@ -19,6 +19,7 @@
     VersionMismatchError,
     DDSArgumentError,
     NoSuchProjectError,
+    CronJobError,
```
Comment (inaod568) [inline dds_web/utils.py:22]:
```suggestion
```

Comment (inaod568) [reply dds_web/utils.py:22]: @valyo produces error

**Comment by inaod568** [comment]
@valyo Seems to be working a lot better now. The common function for bytehours is great. Could you resolve the conflicts and then add some comments for the different parts of the common function? Then I'll do a complete review!

**Comment by valyo** [comment]
While adding comments I realised there is a method timedelta.total_seconds() that returns the whole time difference in seconds

File: dds_web/templates/components/navbar.html [hunk 4]
```diff
@@ -25,12 +25,6 @@
                         Documentation
                     </a>
                 </li>
-                <li class="nav-item">
```
Comment (inaod568) [inline dds_web/templates/components/navbar.html:28]: I think this should still be kept?

**Comment by inaod568** [comment]
> While adding comments I realised there is a method timedelta.total_seconds() that returns the whole time difference in seconds

Ah, nice 👍🏻

File: tests/test_utils.py [hunk 5]
```diff
@@ -892,3 +892,83 @@ def test_bucket_is_valid_ok():
     valid, message = utils.bucket_is_valid(bucket_name="something-.")
     assert valid
     assert message == ""
+
```
Comment (inaod568) [inline tests/test_utils.py:895]: We should add a test to check that the float is ok for a huge file size and a huge number of seconds (> 3 months)

Comment (valyo) [reply tests/test_utils.py:895]: Just tested with:
size_stored: 1000000000000000
hours_stored: 2928.58 (~4 months)

and it seems fine:

```
dds_backend      | [2022-08-30 09:06:11,453] utils [DEBUG] 2928.581792627222
dds_backend      | [2022-08-30 09:06:11,454] scheduled_tasks [INFO] Version: <File Version 1>	Usage: 2.928581792627222e+18
dds_backend      | [2022-08-30 09:06:11,455] scheduled_tasks [INFO] Project someunit00001 byte hours: 2.928581792627222e+18
```
```
┏━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━┳━━━━━━━━━━━┳━━━━━━━━┓
┃ Project ID    ┃ Title          ┃ PI                  ┃ Status      ┃ Last updated                  ┃    Size ┃   Usage ┃      Cost ┃ Access ┃
┡━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━╇━━━━━━━━━━━╇━━━━━━━━┩
│ someunit00001 │ test           │ valyo@me.com        │ In Progress │ Tue, 30 Aug 2022 08:31:17 UTC │ 445.0 B │ 2.9 EBH │ 366.1 Kkr │ ✅     │
│ project_1     │ First Project  │ support@example.com │ In Progress │ Tue, 30 Aug 2022 08:24:05 UTC │   0.0 B │  0.0 BH │    0.0 kr │ ✅     │
│ project_2     │ Second Project │ support@example.com │ In Progress │ Tue, 30 Aug 2022 08:24:05 UTC │   0.0 B │  0.0 BH │    0.0 kr │ ✅     │
```

Comment (valyo) [reply tests/test_utils.py:895]: I also changed test_calculate_bytehours_ok accordingly

File: dds_web/scheduled_tasks.py [hunk 6]
```diff
@@ -230,72 +230,82 @@ def monthly_usage():
...
+            # 1. Get projects where is_active = False
+            # .. a. Check if the versions are all time_deleted == time_invoiced
+            # .. b. Yes --> Set new column to True ("done")
+            scheduler.app.logger.info("Marking projects as 'done'....")
+            for unit, project in page_query(
+                db.session.query(models.Unit, models.Project)
+                .join(models.Project)
+                .filter(models.Project.is_active == False)
+            ):
+                scheduler.app.logger.debug(f"Project: {project}")
```
Comment (inaod568) [inline dds_web/scheduled_tasks.py:251]:
```suggestion
```

File: dds_web/scheduled_tasks.py [hunk 7]
```diff
@@ -230,72 +230,82 @@ def monthly_usage():
...
-                    )
-                    db.session.add(new_record)
-                    db.session.commit()
+                    .count()
+                )
 
+                scheduler.app.logger.debug(f"Number of done versions: {num_done}")
+                scheduler.app.logger.debug(
+                    f"Total number of versions: {len(project.file_versions)}"
+                )
```
Comment (inaod568) [inline dds_web/scheduled_tasks.py:268]:
```suggestion
```

File: dds_web/scheduled_tasks.py [hunk 8]
```diff
@@ -230,72 +230,82 @@ def monthly_usage():
...
+                .join(models.Project)
+                .filter(models.Project.done == False)
+            ):
+                project_byte_hours: int = 0
+                for version in project.file_versions:
+                    if version.time_deleted == version.time_invoiced and [
+                        version.time_deleted,
+                        version.time_invoiced,
+                    ] != [None, None]:
+                        flask.current_app.logger.debug("passing....")
```
Comment (inaod568) [inline dds_web/scheduled_tasks.py:290]:
```suggestion
```

File: dds_web/utils.py [hunk 9]
```diff
@@ -501,3 +501,60 @@ def get_active_motds():
...
+    bytehours = hours_stored * size_bytes
+
+    return bytehours
+
+
+def calculate_version_period_usage(version):
+    bytehours: int = 0
+    if not version.time_deleted and version.time_invoiced:
+        # Existing and invoiced version
+        flask.current_app.logger.debug("1")
```
Comment (inaod568) [inline dds_web/utils.py:526]:
```suggestion
```

File: dds_web/utils.py [hunk 10]
```diff
@@ -501,3 +501,60 @@ def get_active_motds():
...
+        bytehours = calculate_bytehours(
+            minuend=version.time_deleted,
+            subtrahend=version.time_uploaded,
+            size_bytes=version.size_stored,
+        )
+        version.time_invoiced = version.time_deleted
+    elif version.time_deleted != version.time_invoiced:
+        # Version has been deleted after last usage calculation
+        # (if version.time_deleted > version.time_invoiced)
+        flask.current_app.logger.debug("3")
```
Comment (inaod568) [inline dds_web/utils.py:544]:
```suggestion
```

File: dds_web/utils.py [hunk 11]
```diff
@@ -501,3 +501,60 @@ def get_active_motds():
...
+        bytehours = calculate_bytehours(
+            minuend=version.time_deleted,
+            subtrahend=version.time_invoiced,
+            size_bytes=version.size_stored,
+        )
+        version.time_invoiced = version.time_deleted
+    elif not version.time_deleted and not version.time_invoiced:
+        # New version: uploaded after last usage calculation
+        flask.current_app.logger.debug("4")
```
Comment (inaod568) [inline dds_web/utils.py:553]:
```suggestion
```

File: tests/test_utils.py [hunk 12]
```diff
@@ -887,8 +887,91 @@ def test_bucket_is_valid_invalid_suffix():
 
 
 def test_bucket_is_valid_ok():
-    """Test that a bucket name with suffix -s3alias is not valid."""
+    """Test that a bucket name with suffix -s3alias is valid."""
```
Comment (inaod568) [inline tests/test_utils.py:890]:
```suggestion
    """Test that a bucket name with suffix -s3alias is not valid."""
```

File: tests/test_utils.py [hunk 13]
```diff
@@ -887,8 +887,91 @@ def test_bucket_is_valid_invalid_suffix():
...
     assert message == ""
+
+
+# calculate usage
+
+
+def test_calculate_bytehours_ok(client: flask.testing.FlaskClient):
+    """Test that function returns correct value."""
+    minuend: datetime.datetime = datetime.datetime.utcnow()
+    subtrahend = minuend - datetime.timedelta(hours=2928.58)
```
Comment (inaod568) [inline tests/test_utils.py:903]: Could you add a little comment to the side about where this number comes from?

File: tests/test_utils.py [hunk 14]
```diff
@@ -887,8 +887,91 @@ def test_bucket_is_valid_invalid_suffix():
...
     valid, message = utils.bucket_is_valid(bucket_name="something-.")
     assert valid
     assert message == ""
+
+
+# calculate usage
+
+
+def test_calculate_bytehours_ok(client: flask.testing.FlaskClient):
+    """Test that function returns correct value."""
```
Comment (inaod568) [inline tests/test_utils.py:901]:
```suggestion
    """Test that the float and function can handle a huge file stored for 3 months. """
```

File: dds_web/api/project.py [hunk 15]
```diff
@@ -512,10 +512,20 @@ def project_usage(project):
...
 
-            file_hours = (time_deleted - time_uploaded).seconds / (60 * 60)
+            # time_diff_timedelta = time_deleted - time_uploaded
+            # time_diff_days = time_diff_timedelta.days
+            # file_hours = time_diff_days * 24
+            # print(f"Time difference timedelta {time_diff_timedelta}", flush=True)
+            # print(f"Time difference in hours {file_hours}", flush=True)
+            # print(f"Time difference secs {time_diff_timedelta.seconds / (60 * 60)}", flush=True)
+            # print(f"Bytes stored {v.size_stored}", flush=True)
+            # print(f"GBs stored {v.size_stored / 1e9}", flush=True)
```
Comment (inaod568) [inline dds_web/api/project.py:522]:
```suggestion
```

File: dds_web/api/project.py [hunk 16]
```diff
@@ -512,10 +512,20 @@ def project_usage(project):
...
 
             # Calculate BHours
-            bhours += v.size_stored * file_hours
+            # bhours += v.size_stored * file_hours
```
Comment (inaod568) [inline dds_web/api/project.py:525]:
```suggestion
```

File: tests/test_utils.py (hunk 12 above)
Comment (valyo) [reply tests/test_utils.py:890]: I don't get this one, the previous test seem to deal with not valid name and this one asserts valid. It should probably be something like """Test that a bucket name without suffix is valid."""

**Comment by inaod568** [comment]
Approved ✅ 

(Can't approve since I'm the one that created the PR from the beginning)

**Review by valyo — APPROVED** [review]

**Pull request merged**